### PR TITLE
🐛Fix Empty ExtensionElements and Properties

### LINKED ISSUE #794: 🐛When Adding and Removing a Property to/from an Element
#### Describe your issue

When adding and removing a property to/from an element, the element contains an empty executionListener inside the extensionElements.

```
                <bpmn:extensionElements>
                    <camunda:executionListener class="" event="" />
                    <camunda:properties />
                </bpmn:extensionElements>
```

#### How can others reproduce the issue?

- Open a diagram
- Select an element
- export the bpmn
- Add an extension property
- Remove the extension property
- Export the bpmn again
- Compare the two files

#### Possible solution

The full extensionElements part of an element should get removed when the element has no properties.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by alexanderkasten**
:tada: **Thanks for submitting this Pull Request**

Please provide a list including your changes.

To help us to create a meaningful Changelog, we agreed 
to use an emoji-based commit-style. Use the [Emojis](#emojis)
we provided below.

Fill out the following, provide reference to issues
and this PR- we will use this as merge commit message-body
and the title of this PR as commit-headline:

---

**Changes:**

1. Remove empty extensionElements after removing the last extension property of an element
2. Fix the empty properties bug inside an processRef extensionElement (collaboration element)
3. Remove empty extensionElements after removing the last extension property of a participants  processRef of a collaboation
4. Publish a diagramChange event when changing the xml

**Issues:**

Closes #794 

PR: #797 

---

See the [examples](#example) for inspiration.

## How can others test the changes?

- checkout the branch
- `npm run electron-start-dev`
- open a solution
- create a new diagram
- look at the xml and see that there is no extensionElement
- switch to the design view
- select the collaboration or an element
- add some properties
- check the xml again and see them inside an extensionElement
- delete these properties
- check the xml again and see that also the extensionElement has been removed

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 1]
```diff
@@ -96,64 +149,51 @@ export class ProcessSection {
...
 
-    // Set the extension elements of the process reference.
-    this._businessObjInPanel
-        .processRef
-        .extensionElements = extensionElements;
+    const extensionValues: Array<IModdleElement> = [];
+    const properties: Array<IProperty> = [];
+    const propertiesElement: IPropertiesElement = this._moddle.create('camunda:Properties', {values: properties});
+    extensionValues.push(bpmnExecutionListener);
+    extensionValues.push(propertiesElement);
```
Comment (LeoTT) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Would drop the properties variable. 
```ts
    const propertiesElement: IPropertiesElement = this._moddle.create('camunda:Properties', {
      values: [],
    });
```

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 2]
```diff
@@ -47,47 +56,91 @@ export class ProcessSection {
...
+                                                   && extensionValue.values !== undefined
+                                                   && extensionValue.values !== null
+                                                   && extensionValue.values.length !== 0;
+
+          return extensionIsPropertyElement;
+        });
 
-    if (elementHasNoProperties) {
+    const extensionElementHasNoPropertyElement: boolean = extensionsPropertiesElement === undefined
+                                                       || extensionsPropertiesElement === null;
```
Comment (LeoTT) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: `Array.find` delivers undefined, never null.

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 3]
```diff
@@ -47,47 +56,91 @@ export class ProcessSection {
...
+      this._businessObjInPanel
+        .processRef
+        .extensionElements = undefined;
+
+    const removeSingleProperty: (() => Array<IProperty>) = (): Array<IProperty> =>
+      this._propertiesElement
+        .values
+        .splice(index, 1);
+
+    propertyIsLast ? deleteExtensionElements() : removeSingleProperty();
```
Comment (LeoTT) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: See comment above.
```ts
if (propertyIsLast) {
    this._businessObjInPanel.extensionElements = undefined;
} else {
  this._propertiesElement
    .values
    .splice(index, 1);
}
```

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 4]
```diff
@@ -96,64 +149,51 @@ export class ProcessSection {
...
-
-      const extensionIsPropertiesElement: boolean = extensionValue.$type === 'camunda:Properties'
-                                               && extensionValue.values !== undefined
-                                               && extensionValue.values !== null;
+    const propertiesElement: IPropertiesElement  = this._businessObjInPanel
+      .processRef
+      .extensionElements
+      .values
+      .find((extensionValue: IExtensionElement) => {
+      const extensionIsPropertyElement: boolean = extensionValue.$type === 'camunda:Properties';
```
Comment (LeoTT) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: `extensionIsPropertiesElement`

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 5]
```diff
@@ -96,64 +149,51 @@ export class ProcessSection {
...
-    const extensionPropertiesElement: IPropertiesElement = this._moddle.create('camunda:Properties', {values: properties});
+    const emptyProperties: Array<IProperty> = [];
 
-    // Append to the extension elements of the process reference.
-    this._businessObjInPanel
-        .processRef
-        .extensionElements
-        .values
-        .push(extensionPropertiesElement);
+    const createCamundaProperties: (() => IPropertiesElement) = (): IPropertiesElement => this._moddle
```
[the remark this answers is not in the document]
Comment (alexanderkasten) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: @LeoTT @cmg-dev told me I should do it this way in an earlier pull request

[the remark this answers is not in the document]
Comment (Vyperus) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: I like @LeoTT approach. In my eyes it's way more readable.

[the remark this answers is not in the document]
Comment (LeoTT) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Would disregard my comment then, no sense in going back and forth.

But I must admit that I do not see the benefit in adding an arrow functions for that. 
Would love an explanation.

[the remark this answers is not in the document]
Comment (alexanderkasten) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: I agree with you, but to find a good explanation is very hard for me, when there is actually not a use case for the arrow functions.

[the remark this answers is not in the document]
Comment (LeoTT) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Going a little off-topic here.
What I like about arrow functions in situations like these is scoping.
```ts
(() => {
  let a;
  let b;
  // do stuff with a and b
})();
// a and b don't pollute the namespace here
```
But that does not apply in this case.
Would leave it for now and ask @cmg-dev for feedback on this.

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 6]
```diff
@@ -1,3 +1,5 @@
+import {EventAggregator} from 'aurelia-event-aggregator';
+import {inject} from 'aurelia-framework';
```
Comment (Vyperus) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Sort imports, you know how.

Comment (alexanderkasten) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: @Vyperus tslint does not allow me to sort the imports in another order like it is

Comment (alexanderkasten) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: And aurelia-event-aggregator is not installed via our package.json so imo it comes through aurelia => framework. Then this order should be right

Comment (Vyperus) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: I meant you have to add a new line here to seperate both blocks from each other.
Like we do it in all files.

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 7]
```diff
@@ -96,64 +149,50 @@ export class ProcessSection {
...
-
-      const extensionIsPropertiesElement: boolean = extensionValue.$type === 'camunda:Properties'
-                                               && extensionValue.values !== undefined
-                                               && extensionValue.values !== null;
+    const propertiesElement: IPropertiesElement  = this._businessObjInPanel
+      .processRef
+      .extensionElements
+      .values
+      .find((extensionValue: IExtensionElement) => {
+      const extensionIsPropertiesElement: boolean = extensionValue.$type === 'camunda:Properties';
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: indention.

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 8]
```diff
@@ -96,64 +149,50 @@ export class ProcessSection {
...
-    const extensionValues: Array<IModdleElement> = [bpmnExecutionListener, propertiesElement];
-    const extensionElements: IModdleElement = this._moddle.create('bpmn:ExtensionElements', {values: extensionValues});
 
-    // Set the extension elements of the process reference.
-    this._businessObjInPanel
-        .processRef
-        .extensionElements = extensionElements;
+    const extensionValues: Array<IModdleElement> = [];
+    const propertiesElement: IPropertiesElement = this._moddle.create('camunda:Properties', {values: []});
+    extensionValues.push(bpmnExecutionListener);
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Leerzeile darüber einfügen.

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmg-dev) [reply src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: @alexanderkasten is right.

But we can discuss this later. Thanks @alexanderkasten 

@Vyperus I will show you this line after your next vacation and ask you what it does ;)

```ts
const camundaProperties: IPropertiesElement = this._moddle.create('camunda:Properties', {
  values: emptyProperties,
});
```

Brace yourself.

**I think the discussion arises from the awkward lambda syntax of typescript**

We will find a nice way for all of you ;)

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 9]
```diff
@@ -47,47 +59,88 @@ export class ProcessSection {
...
+
+    if (extensionElementHasNoPropertyElement) {
+      return;
+    }
+
+    const properties: Array<IProperty> = extensionsPropertiesElement.values;
     for (const property of properties) {
-      const propertyTypeIsNotCamunda: boolean = property.$type !== 'camunda:Property';
-      if (propertyTypeIsNotCamunda) {
+      if (property.$type !== 'camunda:Property') {
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Why is this change here?!

revert it.

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 10]
```diff
@@ -47,47 +59,88 @@ export class ProcessSection {
...
+        .splice(index, 1);
+    }
+
     this._reloadProperties();
+    this._publishDiagramChange();
   }
 
   public changeName(index: number): void {
     this._propertiesElement.values[index].name = this.newNames[index];
+    this._publishDiagramChange();
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: 👍

File: src/modules/property-panel/indextabs/extensions/sections/process/process.ts [hunk 11]
```diff
@@ -47,47 +59,88 @@ export class ProcessSection {
...
     this.newValues.push('');
 
+    const businessObjectHasNoExtensionElements: boolean = this._businessObjInPanel.processRef.extensionElements === undefined
+                                                       || this._businessObjInPanel.processRef.extensionElements === null;
+    if (businessObjectHasNoExtensionElements) {
+      this._createExtensionElement();
+    }
+
+    this._propertiesElement = this._getPropertiesElement();
+    const propertiesElementIsUndefined: boolean = this._propertiesElement === undefined;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/extensions/sections/process/process.ts]: Leerzeile darüber einfügen.

**Comment by cmg-dev** [comment]
@alexanderkasten War das nicht schon einmal gefixt?!

@robpal94 ist das aufgefallen ;)

**Comment by Vyperus** [comment]
Es gibt tatsächlich zwei Komponenten die ähnliche aber nicht die selben Dinge tun. Daher ist ein ähnlicher Fehler aufgetreten.

**Comment by alexanderkasten** [comment]
@cmg-dev Nein, es war nur für einzelne BPMN-Elemente gefixt und ist es auch immer noch.
Dieser PR fixt das selbe Verhalten, für die Collaboration bzw. für die processRef der Collaboration.
Da werden die extensionElements und Properties an einer anderen stelle im XML gespeichert.

**Pull request merged**